File: typescript-sdk/examples/cli-chat.ts
/**
 * Interactive CLI chat client using the MCPBuilder TypeScript SDK.
 *
 * This example demonstrates a full-featured CLI chat with:
 * - Streaming responses
 * - Tool execution with consent handling
 * - Stop command support
 * - Error handling
 *
 * Usage:
 *   1. Set environment variables:
 *      export MCP_CLIENT_URL=wss://mcp-client.example.com
 *      export PROJECT_TOKEN=your-project-token
 *      export DEPLOYMENT_NAME=your-deployment
 *
 *   2. Run: npx ts-node examples/cli-chat.ts
 */

import * as readline from 'readline';
import { MCPChatClient } from '../src/index.js';

// Configuration from environment
const DEPLOYMENT_NAME = process.env.DEPLOYMENT_NAME || '';
const PROJECT_TOKEN = process.env.PROJECT_TOKEN || '';
const MCP_SERVICE_URL = process.env.MCP_CLIENT_URL || '';
const TIMEZONE = process.env.TIMEZONE || 'Europe/Berlin';

async function main(): Promise<void> {
  if (!DEPLOYMENT_NAME || !PROJECT_TOKEN || !MCP_SERVICE_URL) {
    console.error('❌ Missing required environment variables:');
    console.error('   - DEPLOYMENT_NAME');
    console.error('   - PROJECT_TOKEN');
    console.error('   - MCP_CLIENT_URL');
    process.exit(1);
  }

  // Create client
  const client = new MCPChatClient({
    projectToken: PROJECT_TOKEN,
    deploymentName: DEPLOYMENT_NAME,
    mcpClientUrl: MCP_SERVICE_URL,
    autoReconnect: true,
    maxReconnectAttempts: 5,
  });

  // Track when response is complete
  let responseResolve: (() => void) | null = null;

  // Helper to handle tool consent
  const handleConsent = async (
    consentId: string,
    tool: string,
    args: Record<string, unknown>,
    description?: string
  ): Promise<void> => {
    console.log(`\n⚠️  Tool consent requested: ${tool}`);
    if (description) {
      console.log(`   Description: ${description}`);
    }
    console.log(`   Arguments: ${JSON.stringify(args, null, 2)}`);

    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    const answer = await new Promise<string>((resolve) => {
      rl.question('   Allow? (y/n/a=allow all): ', resolve);
    });
    rl.close();

    const normalizedAnswer = answer.toLowerCase().trim();
    const allow = normalizedAnswer === 'y' || normalizedAnswer === 'a';
    const allowAll = normalizedAnswer === 'a';

    await client.respondConsent(consentId, allow, allowAll);
  };

  // Register callbacks
  client
    .onConnected((sessionId) => console.log(`✅ Connected (session: ${sessionId})`))
    .onToken((text) => process.stdout.write(text))
    .onToolStart((name, input) => {
      console.log(`\n🔧 Calling: ${name}`);
      console.log(`   Input: ${JSON.stringify(input)}`);
    })
    .onToolEnd((name, _output, success, error) => {
      if (success) {
        console.log(`\n✓ ${name} completed`);
      } else {
        console.log(`\n✗ ${name} failed: ${error}`);
      }
    })
    .onToolConsentRequest(handleConsent)
    .onFinal((text, toolCalls) => {
      console.log('\n');
      responseResolve?.();
    })
    .onError((msg) => {
      console.error(`\n❌ Error: ${msg}`);
      responseResolve?.();
    })
    .onDisconnected((reason) => console.log(`\n🔌 Disconnected: ${reason}`));

  console.log(`🔗 Connecting to ${MCP_SERVICE_URL} with deployment '${DEPLOYMENT_NAME}'...`);

  try {
    await client.connect({ timezone: TIMEZONE });
  } catch (err) {
    console.error(`❌ Failed to connect: ${err}`);
    process.exit(1);
  }

  console.log("💬 Type your messages (type 'quit' to exit, 'stop' to cancel):\n");

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  const prompt = (): void => {
    rl.question('You: ', async (input) => {
      const trimmedInput = input.trim();

      if (!trimmedInput) {
        prompt();
        return;
      }

      if (trimmedInput.toLowerCase() === 'quit') {
        console.log('👋 Goodbye!');
        await client.disconnect();
        rl.close();
        process.exit(0);
      }

      if (trimmedInput.toLowerCase() === 'stop') {
        console.log('⏹️  Stopping...');
        await client.stop();
        prompt();
        return;
      }

      process.stdout.write('\nAssistant: ');

      // Create promise to wait for response
      const responseComplete = new Promise<void>((resolve) => {
        responseResolve = resolve;
      });

      // Send message
      await client.sendMessage(trimmedInput);

      // Wait for response with timeout
      const timeout = new Promise<void>((resolve) => setTimeout(resolve, 120000));
      await Promise.race([responseComplete, timeout]);

      prompt();
    });
  };

  prompt();
}

main().catch(console.error);

File: python-sdk/examples/cli_chat.py
#!/usr/bin/env python
"""
Interactive CLI chat client using the MCPBuilder Python SDK.

This example demonstrates a full-featured CLI chat with:
- Streaming responses
- Tool execution with consent handling
- Stop command support
- Error handling

Usage:
    1. Update DEPLOYMENT_NAME and MCP_CLIENT_URL at the top of this file
    2. Run: poetry run python examples/cli_chat.py
"""

import asyncio
import logging
import os
import sys
from pathlib import Path

# ============================================================================
# EXAMPLE CONFIGURATION
# Update these values for your environment
# ============================================================================

DEPLOYMENT_NAME = os.getenv("DEPLOYMENT_NAME", "")      # Your deployment name
PROJECT_TOKEN = os.getenv("PROJECT_TOKEN", "")          # Your project token
MCP_CLIENT_URL = os.getenv("MCP_CLIENT_URL", "")        # MCP client service URL
TIMEZONE = os.getenv("TIMEZONE", "Europe/Berlin")       # Client timezone

# ============================================================================

# Add src to path for development
sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from mcpbuilder import MCPChatClient

# Configure logging
logging.basicConfig(
    level=logging.WARNING,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
)
logger = logging.getLogger(__name__)


async def main():
    """Run interactive chat client."""
    
    # Create client - all configuration passed at initialization
    client = MCPChatClient(
        project_token=PROJECT_TOKEN,
        deployment_name=DEPLOYMENT_NAME,
        mcp_client_url=MCP_CLIENT_URL,
        auto_reconnect=True,
        max_reconnect_attempts=5,
        # cache_history: Server-side history caching (default: False)
        # - False: Client manages history locally (recommended for most cases)
        # - True: Server maintains conversation history across reconnections
    )
    
    # Track when response is complete
    response_complete = asyncio.Event()
    
    # Register callbacks
    client \
        .on_connected(lambda sid: print(f"✅ Connected (session: {sid})")) \
        .on_token(lambda text: print(text, end="", flush=True)) \
        .on_tool_start(on_tool_start) \
        .on_tool_end(on_tool_end) \
        .on_tool_consent_request(lambda cid, tool, args, desc: handle_consent(client, cid, tool, args, desc)) \
        .on_final(lambda text, tool_calls: response_complete.set()) \
        .on_error(lambda msg: print(f"\n❌ Error: {msg}")) \
        .on_disconnected(lambda reason: print(f"\n🔌 Disconnected: {reason}"))
    
    print(f"🔗 Connecting to {MCP_CLIENT_URL} with deployment '{DEPLOYMENT_NAME}'...")
    
    try:
        # Connect - all configuration is set during initialization
        await client.connect(timezone=TIMEZONE)
    except Exception as e:
        print(f"❌ Failed to connect: {e}")
        return
    
    print("💬 Type your messages (type 'quit' to exit, 'stop' to cancel):\n")
    
    try:
        while True:
            # Get user input
            try:
                user_input = await asyncio.get_event_loop().run_in_executor(
                    None, lambda: input("You: ")
                )
            except (EOFError, KeyboardInterrupt):
                print("\n👋 Goodbye!")
                break
            
            user_input = user_input.strip()
            
            if not user_input:
                continue
            
            # Handle special commands
            if user_input.lower() in ["quit", "exit", "/quit", "/exit"]:
                print("👋 Goodbye!")
                break
            
            if user_input.lower() in ["stop", "/stop"]:
                print("⏹️  Stopping current request...")
                await client.stop()
                continue
            
            # Send message and stream response
            print("Agent: ", end="", flush=True)
            response_complete.clear()
            
            try:
                await client.send_message(user_input)
                
                # Wait for response to complete (with timeout)
                try:
                    await asyncio.wait_for(response_complete.wait(), timeout=120.0)
                except asyncio.TimeoutError:
                    print("\n⏱️  Response timeout")
                
                print()  # Newline after response
                
            except asyncio.CancelledError:
                print("\n⏹️  Cancelled")
            except Exception as e:
                print(f"\n❌ Error: {e}")
                logger.exception("Chat error")
    
    finally:
        await client.disconnect()
        print("✅ Disconnected")


def on_tool_start(name: str, input_params: dict):
    """Handle tool start event."""
    print(f"\n🛠️  Using tool: {name}", flush=True)
    if input_params:
        input_str = str(input_params)
        if len(input_str) > 200:
            input_str = input_str[:200] + "..."
        print(f"   Input: {input_str}", flush=True)


def on_tool_end(name: str, output: str, success: bool, error: str | None):
    """Handle tool end event."""
    if success:
        print(f"✅ Tool {name} succeeded", flush=True)
        if output:
            output_str = str(output)
            if len(output_str) > 200:
                output_str = output_str[:200] + "..."
            print(f"   Output: {output_str}", flush=True)
    else:
        print(f"❌ Tool {name} failed", flush=True)
        if error:
            print(f"   Error: {error}", flush=True)


def handle_consent(client: MCPChatClient, consent_id: str, tool: str, args: dict, description: str | None):
    """Handle tool consent request - prompts user for approval."""
    print(f"\n\n⚠️  Tool consent request:", flush=True)
    print(f"   Tool: {tool}", flush=True)
    if description:
        print(f"   Description: {description}", flush=True)
    if args:
        args_str = str(args)
        if len(args_str) > 200:
            args_str = args_str[:200] + "..."
        print(f"   Arguments: {args_str}", flush=True)
    
    try:
        response = input("\n   Allow this tool? (y/n/a=allow all): ").lower().strip()
        
        allow = response in ["y", "yes", "a", "always"]
        allow_all = response in ["a", "always"]
        
        asyncio.create_task(client.respond_consent(consent_id, allow, allow_all))
        
        if allow:
            print("   ✓ Approved", flush=True)
        else:
            print("   ✗ Denied", flush=True)
        
        print("Agent: ", end="", flush=True)
        
    except (EOFError, KeyboardInterrupt):
        asyncio.create_task(client.respond_consent(consent_id, False))
        print("   ✗ Cancelled", flush=True)


if __name__ == "__main__":
    asyncio.run(main())
